activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,161 @@
+    # key format and cache version for cache stores.
+    #
+    # The goal of this class is to provide a composable way
+    # to build keys. You can put an instance of Key
+    # into another instance of Key and it should produce
+    # the same `cache_key` and `cache_version`.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k2 = Key.new(k1)
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method,
+    # key fragments can be added onto an existing key through the
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = Key.new
+    #   key << "foo"
+    #   key << "bar"
+    #   key.cache_key
+    #   # => "foo/bar"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k1 << "bar"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects.
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`,
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.call(key)
+        if key.is_a?(Key)
+          key
+        else
+          self.new(key)
+        end
+      end
```

Comment:
Wondering whether overriding `Key.new` is warranted. Calling `Key.call` suggests I'm somehow invoking Key. What other precedent do we have in Ruby for selectively coercing objects of many kinds to an object of the desired kind? Toplevel methods like `Kernel#Array` and `#Integer` come to mind, and `Hash.[]`.

Reply:
Here's some other options for the `Key.call` method.

```ruby
key = Key.build(name)

key = Key(name)

key = Key[name]
```

I like the `Array()` style call `Key()` the best, since it's behavior that could return a new value or the same value. I wish there was a better english version, but every long attempt I try ends up too unwieldy.

If we stop trying to do the checks and instead simply call `Key.new` each time then we end up losing a signifigant amount of memory savings.

Alternatively we could get rid of this fancy method all together if we instead did this check at each call site:

```ruby
if name.is_a?(Key)
  key = name
else
  key = Key.new(name) 
end
```

That's a little too verbose for my tastes.

Alternatively we could push this behavior to a method inside of `cache.rb` but then we still have a naming issue. Could call it something like `ensure_key()`, `key_or_generate_key()`, `key_or_new()`, etc.

Reply:
I changed it to `Key()` however it's not a kernel method, so the full invocation has to be used `ActiveSupport::Cache::Key("foo")`. What do you think?